pyquil/api/_client.py

```diff
@@ -0,0 +1,241 @@
+
+        assert self._engagement is not None
+
+        return self._rpcq_request(
+            self._engagement.address,
+            method_name,
+            *args,
+            timeout=timeout,
+            auth_config=_to_auth_config(self._engagement.credentials),
+            **kwargs,
+        )
+
+    def _rpcq_request(
+        self,
+        endpoint: str,
+        method_name: str,
+        *args: Any,
+        timeout: Optional[float] = None,
+        auth_config: Optional[ClientAuthConfig] = None,
+        **kwargs: Any,
+    ) -> Any:
+        client = rpcq.Client(endpoint, auth_config=auth_config)
+        response = client.call(method_name, *args, rpc_timeout=timeout, **kwargs)  # type: ignore
+        client.close()  # type: ignore
+        return response
+
+    def reset(self) -> None:
+        """
+        Clears current engagement and reloads configuration (if not overridden in constructor).
+        """
+        self._engagement = None
+        if self._has_default_config:
+            self._config = QCSClientConfiguration.load()
+
+    @property
+    def qvm_url(self) -> str:
+        """
+        QVM URL from client configuration.
+        """
+        return self._config.profile.applications.pyquil.qvm_url
+
+    @property
+    def quilc_url(self) -> str:
+        """
+        Quil compiler URL from client configuration.
+        """
+        return self._config.profile.applications.pyquil.quilc_url
+
+    def qvm_version(self) -> str:
+        """
+        Get QVM version string.
+        """
+        response = self.post_json(self.qvm_url, {"type": "version"})
+        split_version_string = response.text.split()
+        try:
+            qvm_version = split_version_string[0]
+        except ValueError:
+            raise TypeError(f"Malformed version string returned by the QVM: {response.text}")
+        return qvm_version
+
+    @contextmanager
+    def _http_client(self) -> Iterator[httpx.Client]:
+        if self._http is None:
+            with build_sync_client(configuration=self._config) as client:  # type: httpx.Client
+                yield client
+        else:
+            yield self._http
+
+    def _create_engagement(self, processor_id: str) -> EngagementWithCredentials:
+        return cast(
+            EngagementWithCredentials,
+            self.qcs_request(
+                create_engagement,
+                json_body=CreateEngagementRequest(quantum_processor_id=processor_id),
+            ),
+        )
+
+
+def _engagement_valid(processor_id: str, engagement: Optional[EngagementWithCredentials]) -> bool:
+    if engagement is None:
+        return False
+
+    return all(
+        [
+            engagement.credentials.client_public != "",
+            engagement.credentials.client_secret != "",
+            engagement.credentials.server_public != "",
+            engagement.expires_at == "" or parsedate(engagement.expires_at) > datetime.now(tzutc()),
```

Comment:
I don't believe an empty string would be a valid response here. Engagements always expire.

Reply:
OK, will remove